### PR TITLE
refactor: Add motive to Expr.proj

### LINKED ISSUE #2697: (kernel) invalid projection when extracting the proposition from an `Exists` statement.
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
    * Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

The following code explains the issue:

```lean
set_option pp.rawOnError true
/- Error:
(kernel) invalid projection
  h.2
-/
example (h : ∃ _ : Nat, P) : P := h.2

/- Works -/
example (h : ∃ _ : Nat, P) : P := by
  let ⟨_, P⟩ := h
  exact P
```

### Context

The issue is discussed [here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Kernel.20Invalid.20Projection). It seems that the error *makes sense*, but Alex suggested me to open an issue since it's a `kernel` error.

### Versions

The error appears since `Lean (version 4.0.0, commit ec941735c80d, Release)` up til `v4.2.0-rc1`, on a Mac Intel.

### Others

As mentioned above, the error makes sense from a human perspective, so if this `kernel` error is intended, feel free to close this issue :)

### PR DESCRIPTION
**Pull request by digama0**
This is an experiment to add a motive to `Expr.proj`. It is the same as the motive passed to the underlying `casesOn` application, and the kernel checks that it is type correct. This gives the elaborator more flexibility to type projections (see `magical.lean`, also fixes #2697 and supersedes #2700 ), and also (possibly) improves the performance of type inference. But it does add handling of projection motives everywhere, and we need to find out if it is a perf win (or at least neutral).

The plan for this PR was shared ahead of time to @leodemoura . It rewrites `.proj` handling essentially everywhere it is done: in the kernel, in `projections.cpp`, in the app elaborator, in WHNF, and in Meta.Check.

### CODE REVIEW

**Comment by digama0** [comment]
!bench please?

**Comment by Kha** [comment]
!bench

**Comment by leodemoura** [comment]
@digama0 Did it fix the cubic performance issue you mentioned last week?

File: src/Lean/Expr.lean [hunk 1]
```diff
@@ -459,7 +459,7 @@ inductive Expr where
   .proj `Prod 0 a
```
Comment (eric-wieser) [inline src/Lean/Expr.lean:459]: This docstring is outdated

**Comment by digama0** [comment]
> @digama0 Did it fix the cubic performance issue you mentioned last week?

It's unclear. It eliminates this issue for inferType, but when checking the type you still have to do basically the same work and I'm not sure what the distribution is between inferring and checking in the kernel in practice. (The same thing happens in the elaborator.) It does fix a bug where the elaborator was not catching an issue that the kernel discovered but there are less invasive ways to fix that and I'm not sure it is enough motivation on its own. Unfortunately there are a lot of things to do just to get this to compile, you can't really test it without doing the full migration anyway.

Do we have any kernel stress tests? I'm not really sure the existing test suite is sufficient to evaluate kernel performance. I'll see if I can fix the remaining bugs and get a mathlib run (although that will almost certainly require more patching since every expr match is affected).

**Comment by leodemoura** [comment]
@digama0 Thanks for the additional information.

> It's unclear. It eliminates this issue for inferType, but when checking the type you still have to do basically the same work and I'm not sure what the distribution is between inferring and checking in the kernel in practice. 

The kernel performs a lot of checking. Inferring may happen too (e.g., while we are reducing `Eq.rec`), but I would be surprised if it is used more often than checking in the kernel.

> It does fix a bug where the elaborator was not catching an issue that the kernel discovered but there are less invasive ways to fix that and I'm not sure it is enough motivation on its own.

A less invasive solution would be perfect for the bug fix. 

> Do we have any kernel stress tests?

Unfortunately, we don't have. We must need to increase the number of performance tests in core. I have been pushing for more benchmarks and tests in the core repo. It will help us to streamline PRs. 

For this PR, an excellent first step is to construct an example that exhibits the cubic behavior you are concerned about.

**Review by Vtec234 — APPROVED** [review]
Very minor `Lean.Widget` change LGTM!

**Pull request closed without merging**